### PR TITLE
chore: refactor flagd event handling 

### PR DESCRIPTION
**Pull request by Kavindu-Dodan**
## This PR

Completes some of the tasks mentioned at #390 

Introduce clean contracts for flagd evaluator.

- Introduce abstraction layer between flagd OF implementation & GRPC connection handling - `Resolver` interface
- Remove test focused constructors and rely on reflection to inject mocks
- Retry logic is now handled by a dedicated thread & utilize wait notify mechanism. Dummy calls were removed
- Shutdown is propagated to the retry logic handler (thread) to allow proper resource celanup 

Best reviewed either checking out the code OR visiting branch directly on my fork - https://github.com/Kavindu-Dodan/java-sdk-contrib/tree/feat/refactor-flagd/providers/flagd

### CODE REVIEW

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamCallback.java
Comment (Kavindu-Dodan) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamCallback.java:1]: Removed as this was unused

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java
Comment (thisthat) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java]: I think we should also add this to the `onCompleted()` method

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (thisthat) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java]: so we don't confuse this with the wait call reading the logs :) 
```suggestion
                log.debug("failed to sleep while restarting gRPC Event Stream");
```

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (thisthat) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:147]: I have mixed feelings about what to do here. I think it might be worth having a `continue` so we retry until we exhaust the retries

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (liran2000) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java]: Perhaps can use a single threaded (or other) executor, and then shut it down gracefully at shutdown() method ?

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (liran2000) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: For common flow of app shutdown, I guess it can be one of the cases:
1. a graceful shutdown of the thread, it is commonly expected to be interrupted if it is still running - in this case it will print a warning to log if the wait() is interrupted, but is expected as normal positive flow of app shutdown if I understand correctly ?
2. notify() on the object - if I understand correctly, it will proceed to an error ?

There are optional [alternatives](https://stackoverflow.com/questions/4912165/whats-the-best-alternative-to-wait-notify-for-low-level-synchronization#answer-4916691) to wait/notify to consider.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (toddbaert) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java]: I'd still like to see something like this done, if possible. I think that was the meaning of the 3rd checkbox [here](https://github.com/open-feature/java-sdk-contrib/issues/390).

**Comment by thisthat** [comment]
Please, change the wording in the PR description from `Fixes` to `Part of` or something similar. Otherwise the original issue will be closed when not all the points are addressed

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java
Comment (liran2000) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java]: If this can be called from multiple threads, and the other thread which wait on this object is during wait(), then this here may be stuck on synchronized ?
Is it possible ?

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (liran2000) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:143]: Can consider adding [Jitter](https://aws.amazon.com/builders-library/timeouts-retries-and-backoff-with-jitter/).

**Review by liran2000 — COMMENTED** [review]
Hi, added some comments, hope it's ok.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java]: @thisthat good point. `onComplete` is a last call similar to `onError`. So we must reinitialize the listener

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java]: @liran2000 we do not have multi-thread model here. Let me try to explain, 

We start a thread to handle the stream observer [^1], this is the only extra thread that we create. This thread handles a retry mechanism [^2] and executes an asynchronous `EventStreamObserver`.

`EventStreamObserver` is an implementation of grpc `StreamObserver`. `StreamObserver` implementation is such that `onError` & `onCompleted` calls are called once, and `no further calls to any method are allowed` [^3]

We `notify` waiting caller when dealing with `onError` & `onCompleted`. Given these are the last calls, only the freshly initialized  `EventStreamObserver` can call the `notify` if that too face with an error. 

[^1]: https://github.com/Kavindu-Dodan/java-sdk-contrib/blob/feat/refactor-flagd/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L83

[^2]: https://github.com/Kavindu-Dodan/java-sdk-contrib/blob/feat/refactor-flagd/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L119

[^3]: https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/StreamObserver.html#onError(java.lang.Throwable)

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:147]: fair point 🤔  I will update this to continue

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (toddbaert) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java]: > Perhaps can use a single threaded (or other) executor, and then shut it down gracefully at shutdown() method ?

Ya, I generally like to use single thread executors instead of threads directly.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (toddbaert) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: I think this wait/notify can be expanded to prevent us from having to do a dummy evaluation, as I described [here](https://github.com/open-feature/java-sdk-contrib/pull/378#discussion_r1281102577).

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:147]: Updated with the latest commit

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java]: I am revisiting the loigc. will see if we can get rid of dummy call

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java]: Unfortunately, we need some sort of a concurrent mechanism to execute the connection retry logic [^1]. This is to support, busy wait (sleep) and the loop for retry counter.

However, I agree with the point of graceful shutdown (thread termination on shutdown invocation). I will work on this improvement 

[^1]: https://github.com/Kavindu-Dodan/java-sdk-contrib/blob/feat/refactor-flagd/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L115-L144

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:147]: I had to revert this with graceful shutdown support [^1] . This is needed to clean up this thread for provider shutdown. 

[^1]: https://github.com/Kavindu-Dodan/java-sdk-contrib/blob/feat/refactor-flagd/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L84-L85

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: @liran2000  good points.

I have added thread cleanup backed by `thread.interrupt` [^1]. And downgraded the logs to debug level. From OpenFeature perspective, API provides shutdown support, which flagd implements. And the new logic respects this shutdown call by accepting an interrupt (then log and exit the thread).

 Once the thread is interrupted and lead to a shutdown, `notify` is no-op and shall not cause any error. Then the GC should take care of the rest of the cleanup  



[^1]: https://github.com/Kavindu-Dodan/java-sdk-contrib/blob/feat/refactor-flagd/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L84-L85

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: Thanks for the follow-up. Completed the cleanup with this commit - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/cbaa368a0867f5de6ee23d089aeeae6b792e3eb4

We no longer rely on the dummy call

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:143]: > If this can be called from multiple threads, and the other thread which wait on this object is during wait(), then this here may be stuck on synchronized ?
Is it possible ?

As explained here [^1], we do not use multiple threads. We only have 1-1 relationship from caller to async stream handler.

Given sync block is the standard way to deal with `wait` & `notify` [^2], we shall not face with deadlocks. 

Regarding adding jitter, I think this can be considered a follow-up 

[^1]: https://github.com/open-feature/java-sdk-contrib/pull/398#discussion_r1296076334
[^2]: https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#notify--

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (liran2000) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: thanks, what about the case of handling notify from onCompleted, is this a valid scenario ? as I see in this case, wait() returns without exception, the code then proceeds to sleep() - not sure it this will cause immediate interrupted exception  and return or will finish the sleep and will proceed to unexpected error.
Maybe add log info/debug and return immediately after wait() can make it clear ? what do you think ?
If this specific case is covered by unit test it can be helpful.

**Comment by Kavindu-Dodan** [comment]
> Please, change the wording in the PR description from `Fixes` to `Part of` or something similar. Otherwise the original issue will be closed when not all the points are addressed

Done. However, we can mark `Use always OTel interceptor` can be considered done with the new packaging structure

**Comment by Kavindu-Dodan** [comment]
@thisthat @toddbaert @liran2000  thank you for the reviews 🙌 

I addressed all your concerns. As summary,

- Retry logic now solely rely on the eventing mechanism (no dummy calls)
- Graceful shutdowns for event listener implemented - proper resource cleanup 
- Justification on thread usage along with proper log levels and comments where necessary

**Comment by liran2000** [comment]
@Kavindu-Dodan thanks :) had one [question](https://github.com/open-feature/java-sdk-contrib/pull/398#discussion_r1296273468) above.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: `onCompleted` means a successful stream completion. If flagd provide stream closure without error, then this will handle this scenario. 

We need to handle this as this is also a final call similar to `onError`

When this (either `onError` or `onComplete`) happens, we disconnect from flagd. This can be resulted from a restart (ex:- new version rollout), OOMKILL or similar. From SDK point of view, we attempt to reconnect with backoff. This is why we have a sleep and continue with a new stream watcher connection

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: I will try to add some unit tests to cover this scenario :)

**Comment by toddbaert** [comment]
> @thisthat @toddbaert @liran2000 thank you for the reviews 🙌
> 
> I addressed all your concerns. As summary,
> 
> * Retry logic now solely rely on the eventing mechanism (no dummy calls)
> * Graceful shutdowns for event listener implemented - proper resource cleanup
> * Justification on thread usage along with proper log levels and comments where necessary

Amazing! I will re-review tomorrow.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:129]: Added unit tests to verify retry attempts - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/1af6f4ce52f326bb12482da906957013acade150#diff-0dff792882d8d7962486094010e2c59c4210efa38a4b0f7ee4d5f964934eb3c4

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (toddbaert) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:79]: This reads much simpler, but we aren't blocking until we actually get a good connection, like I mentioned [here](https://github.com/open-feature/java-sdk-contrib/pull/378#discussion_r1281102577).

This means that flagd can't be reached, we'll see a READY from the SDK (because initialize returns immediately) and then after that get an error from the thread. Unless I'm missing something, this could be fixed simply by waiting on the `sync` here and notifying on successful connect to unblock.

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:79]: > This means that flagd can't be reached, we'll see a READY from the SDK

This is incorrect. Let me try to explain.

Flagd provider is an `EventProvider` with a default status of `NOT_READY`. WIth the current refactoring, status is fully maintained by event subscriptions, observed concurrently through this logic [^1]. The actual event subscriptions are handled by `EventStreamObserver` [^2].

Regarding initialization returning immediately, given that we have start status of `NOT_READY`, we are sticking to this status till we perform a callback from the event observer. And this happens fast as soon as event observer attempt to connect with flagd. And there are two possibilities that happens at `EventStreamObserver`

- Successful connection - results in a `ProviderReady` sent by flagd as the initial handshake

This makes `EventStreamObserver` to callback with ready status, making provider ready

- Connection failure - results in invoking `onError`

This makes `EventStreamObserver` to callback with`Error` status, making provider status to `ERROR`

Hope this helps to clarify your doubt. But please let me know if I am not seeing something here :) 


[^1]: https://github.com/open-feature/java-sdk-contrib/blob/1af6f4ce52f326bb12482da906957013acade150/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L114-L144
[^2]: https://github.com/open-feature/java-sdk-contrib/blob/main/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java#L19

Comment (toddbaert) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:79]: > Regarding initialization returning immediately, given that we have start status of NOT_READY, we are sticking to this status till we perform a callback from the event observer. And this happens fast as soon as event observer attempt to connect with flagd. And there are two possibilities that happens at EventStreamObserver

Regardless of this, when this `initialize` returns successfully, the SDK fires a READY event. This is what I mean when I said "we'll see a READY from the SDK".

The SDK expects the provider to be in a usable state immediately after the return of `initialize`.

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:79]: Addressed this with the latest commit - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/a21363ead7e869ab9550d180c385dcc36744ea0b

Now we perform a busy wait for a successful connection. This is done with a countdown similar to what you proposed - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/a21363ead7e869ab9550d180c385dcc36744ea0b#diff-67d657d2b09b8caa955560fe7640113ec2bb5528db86e0f96dc15b83e86c9627R172-R182

please check and let me know your thoughts :)

Comment (toddbaert) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:79]: looks good!

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java
Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java:143]: Given that adding jitter is a simple change, I added this to our logic. See commit - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/d28bdcd679fef903566697677b7d5a8797417c18

**Review by toddbaert — APPROVED** [review]
Manually tested this (though we do have the e2e tests now :sunglasses:). LGTM now. Thanks for the improvements.

**Comment by Kavindu-Dodan** [comment]
> Manually tested this (though we do have the e2e tests now 😎). LGTM now. Thanks for the improvements.

Thank you 🙏 I also did several manual validations against a local flagd instance. But yeah, e2e tests to helps

**Pull request merged**
